picard/file.py

```diff
@@ -280,25 +301,25 @@ def _save_and_rename(self, old_filename, metadata):
                 if not tail:
                     head, tail = os.path.split(head)
                 while head and tail:
-                    try:
-                        self._rmdir(head)
-                    except BaseException:
-                        break
+                    self._rmdir(head)
                     head, tail = os.path.split(head)
-            except EnvironmentError:
-                pass
+            except OSError as why:
+                log.warning("Error removing directory: %s", why)
+            except SkipRemoveDir as why:
+                log.debug("Not removing empty directory: %s", why)
         # Save cover art images
         if config.setting["save_images_to_files"]:
             self._save_images(os.path.dirname(new_filename), metadata)
         return new_filename
 
     @staticmethod
     def _rmdir(path):
-        junk_files = (".DS_Store", "desktop.ini", "Desktop.ini", "Thumbs.db")
-        if not set(os.listdir(path)) - set(junk_files):
-            shutil.rmtree(path, False)
+        if os.path.realpath(path) in PROTECTED_DIRECTORIES:
+            raise SkipRemoveDir('%s is a protected directory' % path)
+        elif set(os.listdir(path)) - JUNK_FILES:
```

Comment:
I wonder about another approach:

- try to os.rmdir(path): if it fails, either it's because of permissions or the like, then bail out, else it may fail because it's non-empty, in this case remove junk files from it, and try again, if it fails again, bail out

This way we don't have to calll listdir() at all, and it's faster when no junk files exist.

What do you think?

Reply:
>     * in this case remove junk files from it, and try again, if it fails again, bail out

We must not do this, the "junk" files are not necessarily there to be removed at will, we should only remove them if we indeed remove the directory. Example is the `desktop.ini` files on windows: Those control how the directory is displayed in Windows Explorer. If we just remove them in the hope they are the only file in the dir we might break user configurations. They just don't block the directory from being considered empty.

So in case the `os.rmdir` would fail we then would still need to call `os.listdir` to figure out whether the junk files are really all that is there, which wouldn't help us much.

But I agree that this needs better testing, I'm going to factor out this code and add proper tests.

Reply:
Ok, make sense

Reply:
I moved the empty dir detection and removal into a separate module and addes tests, please take a look.